### DISCUSSION TITLE
Domain entry special cases

### DISCUSSION THREAD

**Discussion by w3suli** [fyne-io/fyne#4891] [Q&A] [answered]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

I am creating a domain input field in the fyne graphical environment.
The basic input is easy to filter, but there were problems.
The domain name cannot contain two dots in a row.
There cannot be a hyphen before or after the period.
My questions:
If, while typing, the user moves the cursor in the already typed text. Is it possible to scan the character before the cursor and the character after the cursor?
If the user selects a piece of text, can the character before and after the selection be read?
The goal is to prevent the user from writing or pasting text that invalidates the already entered domain name.

![cursor-position](https://github.com/fyne-io/fyne/assets/20889555/1a83ad82-e12d-437f-b4be-b69b8878f863)
![select-position](https://github.com/fyne-io/fyne/assets/20889555/b3fd8ade-06de-43ec-828d-7da73270ff1a)

### Describe the solution you'd like to see.

I would like to detect when the user clicks on the entry and then request the characters before and after the cursor in the program.
I would like to detect when the user clicks in the entry and selects a text section and then request the characters before and after the selection in the program.

**Comment by andydotxyz** [comment]
It sounds like a case for Validation.

Otherwise an extended Entry where you intercept the TypedRune and paste functionality to do as you desire.

**Reply by w3suli** [reply]
Domain entry would be a further development of the [Numerical Entry](https://docs.fyne.io/extend/numerical-entry) solution.
The challenge is that it cannot be filtered only at the character level.
There are forbidden character combinations. For example: .. or -. and .-
As long as the user continues to write the text, these compositions can be easily filtered.

If the user moves the cursor, the characters before and after the cursor cannot be read, so it is not possible to filter the formation of forbidden character combinations.
Similarly, if the user selects a section in an existing text, a similar problem occurs. In order to be able to filter the possibility of writing or pasting, the character preceding and following the given position would be needed.

I would like to prevent inappropriate characters from being entered into the entry field.

**Reply by andydotxyz** [reply]
can't you just access Entry.Text to get the full text and work with that?

**Reply by w3suli** [reply]
If there isn't a better solution, I thought about it too, but then the new character or part of the text would appear and it would be withdrawn afterwards.
It would be better if it were checked before publication.

In the case of the domain entry, one character before and after the cursor position or selection position is required. In other cases, it can be several characters or even the entire text to the left and right of the cursor or selection. There may be more complicated composition analyzes and it may happen that the user sees the typed or pasted text appear and disappear, which is not a nice solution.
That's why I want to analyze the errors before the release, if possible.

**Reply by dweymouth** [reply]
If you extend Entry and override the TypedKey function, you will intercept every character and can choose whether to pass it along to the underlying widget.Entry. You can check the current text (entry.Text) and the cursor column, and you should be able to write whatever logic you need to accept or reject that character

**Reply by w3suli** [reply]
I solved the first problem, but I can't manage the second one.
The start and stop values ​​also included in the function below are required, which I cannot achieve. This is what I need: `start, stop := e.selection()`. However, I cannot access `e.selection()`.


```
func (e *Entry) SelectedText() string {
	e.propertyLock.RLock()
	defer e.propertyLock.RUnlock()

	if !e.selecting {
		return ""
	}

	start, stop := e.selection()
	if start == stop {
		return ""
	}
	r := ([]rune)(e.textProvider().String())
	return string(r[start:stop])
}
```

**Reply by andydotxyz** [reply]
The cursor position will be at the end of the selection, and the start will be `len(e.SelectedText())` runes earlier.

**Reply by w3suli** [reply]
Unfortunately not. That's what I thought at first too.
Text can be selected not only from left to right, but also from right to left.
Also, there may be repeated sections of text, and in this case, even scanning the text on both sides of the cursor and comparing it with the content of the selected section does not guarantee the correct position.

**Reply by andydotxyz** [reply]
> Text can be selected not only from left to right, but also from right to left.

Hmm I see. I will have to think about that - but you could keep track of the cursor position when selection is set, then follow it for changes. Not ideal but that is possible with existing API...

> Also, there may be repeated sections of text, and in this case, even scanning the text on both sides of the cursor and comparing it with the content of the selected section does not guarantee the correct position.

Absolutely - that is not a suitable solution.

**Comment by w3suli** [comment]
To extract the characters before and after the cursor position, `e.CursorColumn` can be used well (`e.Text[e.CursorColumn]`), so part of the problem is solved.
However, I do not have access to the starting and ending positions of the selection. Would there be a way to query the start and end position of the marking?

**Comment by andydotxyz** [comment]
```go
func main() {
	a := app.New()
	w := a.NewWindow("Hello")

	hello := widget.NewLabel("")
	ent := widget.NewEntry()
	selstart := -1
	ent.OnCursorChanged = func() {
		pos := ent.CursorColumn
		text := ent.SelectedText()
		if text == "" {
			hello.SetText("No selection")
			selstart = -1
			return
		}
		pos2 := 0
		if selstart == -1 {
			selstart = ent.CursorColumn
		}

		if ent.CursorColumn >= selstart {
			pos = ent.CursorColumn - len(text)
		}
		pos2 = pos + len(text)

		info := fmt.Sprintf("Selection \"%s\", range %d:%d", text, pos, pos2)
		hello.SetText(info)
	}
	w.SetContent(container.NewVBox(
		hello,
		ent,
	))

	w.ShowAndRun()
}
```

**Reply by w3suli** [reply]
Partly true. I didn't think of that, thanks!
There is one problem with the solution:
If the user clicks into the text and then presses the Shift key to make the selection, strange results can occur.
Example:
0123456789
In the number line above, by placing the cursor after the number eight, then pressing the Shift key and clicking behind the number three, the program returns the value -2:3 (the correct value would be 3:8).
Or by clicking after the nine and pressing Shift, then clicking after the four, the result will be -1:4 (the correct one would be 4:9).
Or clicking after the seven and holding down the shift key and clicking after the single will result in -5:1 instead of a1:7.

**Reply by andydotxyz** [reply, marked as answer]
```go
// Package main loads a very basic Hello World graphical application.
package main

import (
	"fmt"

	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Hello")

	hello := widget.NewLabel("")
	ent := widget.NewEntry()
	selstart := -1
	lastCol := 0
	ent.OnCursorChanged = func() {
		pos := ent.CursorColumn
		text := ent.SelectedText()
		if text == "" {
			hello.SetText("No selection")
			lastCol = pos
			selstart = -1
			return
		}
		pos2 := 0
		if selstart == -1 {
			selstart = lastCol
		}

		if ent.CursorColumn >= selstart {
			pos = ent.CursorColumn - len(text)
		}
		pos2 = pos + len(text)

		info := fmt.Sprintf("Selection \"%s\", range %d:%d", text, pos, pos2)
		hello.SetText(info)
	}
	w.SetContent(container.NewVBox(
		hello,
		ent,
	))

	w.ShowAndRun()
}

```

**Reply by w3suli** [reply]
Looks good, thank you very much!
